design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+# Improvements to virtual chunks handling
+
+## What we need
+
+1. Virtual chunks can be stored in all supported object stores
+2. Multiple credentials sets for virtual chunk resolution
+3. Optimized manifests where the same prefix doesn't need to be repeated for every virtual chunk
+4. Be able to know all the "places" pointed by a repo, without having to load every manifest in every version
+5. Object level change detection (not serving a virtual chunk if the containing file changed)
+
+## Discussion
+
+### Chunks url styles
+
+* Many chunks from a few objects, the prefix can be the file
+  * Example: `s3://some-bucket/some-prefix/some-file.nc` is pointed to by 10k chunks.
+  * Having between 10 and 10k chunks from the same file is not unusual
+* Virtual zarr: each chunk comes from a different object, the prefix is not the object, we still need an extra string
+  * Example: `s3://some-bucket/some-prefix/c/0/0/1` is pointed to by only 1 chunk. But there are many objects like it.
+  * There could be tens of millions of chunks (so different urls)
+
+### Prefixes
+
+Storing the full URL for each chunk duplicates a lot of data in the manifest, making it large. The idea is to try to extract common _prefixes_ and identify them with a short id. Then the manifest only needs to identify the prefix id + maybe an extra string + length + offset.
+
+In the first url style above, the prefix could be the full file, `s3://some-bucket/some-prefix/some-file.nc`. In the second example, there are options for the prefix, it could be `s3://some-bucket/some-prefix/c/0/0/1`, or `s3://some-bucket/some-prefix/c/0/0` or `s3://some-bucket/some-prefix/c/0` or `s3://some-bucket/some-prefix/c`, with different numbers of chunks under it in each case.
+
+Intelligence is needed to optimize what is consider a good prefix. If done badly, we could have:
```

Comment:
Is intelligence really needed? Or would an off-the shelf compression algorithm handle it just fine?

I decided to test this with my favorite little storage library

```python
import zarr
import numpy as np

refix = 's3://some-bucket/some-prefix/c/0/'
paths = np.array([str(i) for i in range(10000)])
full_paths = np.array([prefix + p for p in paths])

store = zarr.MemoryStore()
group = zarr.group(store=store)

path_array = group.create("paths", shape=len(paths), chunks=len(paths), dtype=str, compressor=zarr.Zstd())
path_array[:] = paths

full_path_array = group.create("full_paths", shape=len(paths), chunks=len(paths), dtype=str, compressor=zarr.Zstd())
full_path_array[:] = full_paths

print(path_array.nbytes_stored)
print(full_path_array.nbytes_stored)
# 6579
# 7048
```

So storing the full paths and then using a zstd compressor on the raw bytes creates only 7% more data than just storing the last part of the path.

This is what compression algorithms do: find redundant patterns in data. Should we really spend time trying to implement our own compression scheme for prefixes?

Reply:
This is my failure to clarify that we were looking for an in-memory optimization. Currently Icechunk lacks a layer of separation between the in memory and on-disk datastructures (that will have to come soon) which made things even more confusing.